libz-rs-sys/src/gz.rs

```diff
@@ -782,4 +846,47 @@ mod tests {
         // gzclose should return an error because the fd is invalid.
         assert_eq!(unsafe { gzclose(handle) }, Z_ERRNO);
     }
+
+    #[test]
+    fn test_gzclearerr_and_gzeof() {
+        // gzclearerr on a null file handle should return quietly.
+        unsafe { gzclearerr(ptr::null_mut()) };
+
+        // gzeof on a null file handle should return false.
+        assert_eq!(unsafe { gzeof(ptr::null_mut()) }, 0);
+
+        // Open a gzip stream with an invalid file handle. Initially, no error
+        // status should be set.
+        let handle = unsafe { gzdopen(-2, c!(b"r\0")) };
+        assert!(!handle.is_null());
+        assert_eq!(unsafe { gzeof(handle) }, 0);
+
+        // gzclearerr should reset the eof and past flags.
+        let state = (unsafe { handle.cast::<GzState>().as_mut() }).unwrap();
+        state.eof = true;
+        assert_eq!(unsafe { gzeof(handle) }, 0);
+        state.past = true;
+        assert_eq!(unsafe { gzeof(handle) }, 1);
+        unsafe { gzclearerr(handle) };
+        assert!(!state.eof);
```

Comment:
The Miri test run flags this line as an error because of a "foreign write" to state.eof by gzclearerr. I think that might be because the pointer aliasing and downcasting here has confused Miri, but it could use a second set of eyes.

Reply:
the aliasing is in violation of rust's rules: `state` is a mutable reference, which implies exclusive access, but its value is read via `handle`. You need to use one or the other, and I suspect here it's easiest to just never create the mutable reference there. I'll post a suggestion

Reply:
oh and because miri is kind of slow, smaller tests are better (they can be evaluated in parallel). Does splitting this out into a test for gzeof and gzclearerror have big downsides?